src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,479 @@
+/-
+Copyright (c) 2021 Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kexing Ying
+-/
+import measure_theory.vector_measure
+import order.symm_diff
+
+/-!
+# Hahn decomposition
+
+This file prove the Hahn decomposition theorem (signed version). The Hahn decomposition theorem
+states that, given a signed measure `s`, there exist complement, measurable sets `i` and `j`,
+such that `i` is positive and `j` is negative with repsect to `s`; that is, `s` restricted on `i`
+is non-negative and `s` restricted on `j` is non-positive.
+
+The Hahn decomposition theorem leads to many other results in measure theory, most notably,
+the Jordan decomposition theorem, the Lebesgue decomposition theorem and the Radon-Nikodym theorem.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_subset_restrict_nonpos` : A measurable set of negative measure contains
+  a negative subset.
+
+## Notation
+
+We use the notations `0 ≤[i] s` and `s ≤[i] 0` to denote the usual definitions of a set `i`
+being positive/negative with respect to the signed measure `s`.
+
+## Tags
+
+Hahn decomposition theorem
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal measure_theory
+
+variables {α β : Type*} [measurable_space α]
+variables {M : Type*} [add_comm_monoid M] [topological_space M] [ordered_add_comm_monoid M]
+
+namespace measure_theory
+
+namespace signed_measure
+
+open filter vector_measure
+
+variables {s : signed_measure α} {i j : set α}
+
+section exists_subset_restrict_nonpos
+
+/-! ### exists_subset_restrict_nonpos
+
+In this section we will prove that a set `i` whose measure is negative contains a negative subset
+`j` with respect to the signed measure `s` (i.e. `s ≤[j] 0`), whose measure is negative. This lemma
+is used to prove the Hahn decomposition theorem.
+
+To prove this lemma, we will construct a sequence of measurable sets $(A_n)_{n \in \mathbb{N}}$,
+such that, for all $n$, $A_{n + 1} \subseteq i \setminus \bigcup_{k \le n} A_k$ and
+$\frac{1}{a_n + 1} < s(A_{n + 1})$ for some sequence $a_n \to \infty$.
+
+This sequence of sets does not necessarily exist. However, if this sequence terminates; that is,
+there does not exists any sets satisfying the property, the last $A_n$ will be a negative subset
+of negative measure, hence proving our claim.
+
+In the case that the sequence does not terminate, it is easy to see that
+$i \setminus \bigcup_{k = 0}^\infty A_k$ is the required negative set.
```

Comment:
This is slightly misleading, no? If I understand correctly, for the proof to work, you have to choose carefully the set `A_n`: its measure should be almost maximal among sets with positive measure, which you haven't mentioned in the above sketch.

To make this clear, I'd probably replace line 60 with:
 `s (A_{n+1})` is close to maximal among sets in `i \setminus \bigcup_{k \le n} A_k` with positive measure
and I'd remove line 75 and the second half of line 74.